### PR TITLE
Small change to how maps are serialized in the protobuf wire format

### PR DESCRIPTION
**Pull request by wjtje**
Previously, maps were encoded where the field number of the map was written once with the length of all elements in the map. Now the map field number is repeated for each entry, which more closely follows the protobuf syntax.

```proto
message MapFieldEntry {
  key_type key = 1;
  value_type value = 2;
}

repeated MapFieldEntry map_field = N;
```

---

Before:

```
1: {
  1: 1
  2: 2
  1: 2
  2: 3
}
```

After:

```
1: {
  1: 1
  2: 2
}
1: {
  1: 2
  2: 3
}
```

### CODE REVIEW

File: test/pb.cpp [hunk 1]
```diff
@@ -1740,7 +1740,7 @@ TEST_CASE("protobuf")
         {
             CHECK(pb_serialize<mode>(std::map<int32_t, int32_t>{{1, 2}}) == "\x0a\x04\x08\x01\x10\x02");
             CHECK(pb_serialize<mode>(std::map<int32_t, int32_t>{{1, 2}, {2, 3}}) ==
-                  "\x0a\x08\x08\x01\x10\x02\x08\x02\x10\x03");
+                  "\x0a\x04\x08\x01\x10\x02\x0a\x04\x08\x02\x10\x03");
```
Comment (tonda-kriz) [inline test/pb.cpp]: Please use `pb_json_test(...)` to test all variants (pb/json/serialize/deserialize) because if `pb::serialize` was wrong, `pb::deserialize` will most likely be also wrong...

**Review by tonda-kriz — CHANGES_REQUESTED** [review]
Hi, you are right, according to https://protobuf.com/docs/language-spec#maps, maps are encoded as repeated structs and not as repeated packed. Thank you for the PR.

File: test/pb.cpp (hunk 1 above)
Comment (wjtje) [reply test/pb.cpp]: I switched to `pb_json_test` for the `string/string` test but could not get it working for the tests with `int32`; I kept running into an "invalid wire type" exception and could not figure out why.

I did find a small issue where the key of the map was empty when serializing it into JSON.

Comment (tonda-kriz) [reply test/pb.cpp]: Empty map string key? Interesting edge case...

The `void deserialize(auto &stream, std::map<keyT, valueT> &value, wire_type type)` is wrong, its using the wire type from the function argument (`type`), which is always `length_delimited` (same as for string/messages), that's why the `string/string` is working and nothing else. The function needs to use `field_type` instead of `type`, the following diff should fix it  

```
diff --git a/include/spb/pb/deserialize.hpp b/include/spb/pb/deserialize.hpp
index 6e03865..42d5f63 100644
--- a/include/spb/pb/deserialize.hpp
+++ b/include/spb/pb/deserialize.hpp
@@ -641,10 +641,9 @@ inline void deserialize(auto &stream, std::map<keyT, valueT> &value, wire_type t
         switch (field_number)
         {
         case 1:
-            check_wire_type_or_throw(field_type, type);
             if constexpr (std::is_integral_v<keyT>)
             {
-                deserialize<key_encoder>(stream, pair.first, type);
+                deserialize<key_encoder>(stream, pair.first, field_type);
             }
             else
             {
@@ -652,21 +651,20 @@ inline void deserialize(auto &stream, std::map<keyT, valueT> &value, wire_type t
                 {
                     const auto size = read_varint<uint32_t>(stream);
                     auto substream = stream.sub_stream(size);
-                    deserialize<key_encoder>(substream, pair.first, type);
+                    deserialize<key_encoder>(substream, pair.first, field_type);
                     check_if_empty_or_throw(substream);
                 }
                 else
                 {
-                    deserialize<key_encoder>(stream, pair.first, type);
+                    deserialize<key_encoder>(stream, pair.first, field_type);
                 }
             }
             key_defined = true;
             break;
         case 2:
-            check_wire_type_or_throw(field_type, type);
             if constexpr (spb::detail::proto_field_number<valueT>)
             {
-                deserialize<value_encoder>(stream, pair.second, type);
+                deserialize<value_encoder>(stream, pair.second, field_type);
             }
             else
             {
@@ -674,7 +672,7 @@ inline void deserialize(auto &stream, std::map<keyT, valueT> &value, wire_type t
                 {
                     const auto size = read_varint<uint32_t>(stream);
                     auto substream = stream.sub_stream(size);
-                    deserialize<value_encoder>(substream, pair.second, type);
+                    deserialize<value_encoder>(substream, pair.second, field_type);
                     check_if_empty_or_throw(substream);
                 }
                 else [[unlikely]]

```

**Review by tonda-kriz — APPROVED** [review]
Looks good, Thank you
